networkit/cpp/io/NetworkitBinaryWriter.cpp

```diff
@@ -24,8 +24,10 @@ void NetworkitBinaryWriter::write(const Graph &G, const std::string &path) {
 	Aux::enforceOpened(outfile);
 	nkbg::WEIGHT_FORMAT weightFormat;
 
-	auto detectWeightsType = [&] () {
-		weightFormat = nkbg::WEIGHT_FORMAT::VARINT;
+	auto detectWeightsType = [&] () -> nkbg::WEIGHT_FORMAT {
+		if(!G.isWeighted())
+			return nkbg::WEIGHT_FORMAT::NONE;
+
 		bool isUnsigned = true;
 		bool fitsIntoInt64 = true;
 		bool fitsIntoFloat = true;
```

Comment:
While you didn't touch this code: Shouldn't this be `G.forEdges` rather than `G.forNodes .. G.forNeighborsOf` ? It seems to be the faster and more idiomatic choice.

Reply:
Also we can make the loop practically branch-free: https://godbolt.org/z/S8WwPd

Reply:
I agree about the `G.forEdges`.

About the branch-free version: It's not immediately clear that the branch-free code is faster. You trade a control flow dependency for a data flow dependency.

Reply:
Though that is clearly not enough, but I did some benchmarks with the writer in both loop variations for synthetic networks (order of 10 million nodes, random negative and positive float-values). For this, the results indicate a slight advantage in the "original" version with branches